libs/wire-api/src/Wire/API/OAuth.hs

```diff
@@ -0,0 +1,596 @@
+      redirectUrlDescription = description ?~ "The URL to redirect to after the user has authorized the application."
+
+newtype OAuthClientPlainTextSecret = OAuthClientPlainTextSecret {unOAuthClientPlainTextSecret :: AsciiBase16}
+  deriving (Eq, Generic)
+  deriving (A.ToJSON, A.FromJSON, S.ToSchema) via (Schema OAuthClientPlainTextSecret)
+
+instance Show OAuthClientPlainTextSecret where
+  show _ = "<OAuthClientPlainTextSecret>"
+
+instance ToSchema OAuthClientPlainTextSecret where
+  schema = (toText . unOAuthClientPlainTextSecret) .= parsedText "OAuthClientPlainTextSecret" (fmap OAuthClientPlainTextSecret . validateBase16)
+
+instance FromHttpApiData OAuthClientPlainTextSecret where
+  parseQueryParam = bimap cs OAuthClientPlainTextSecret . validateBase16 . cs
+
+instance ToHttpApiData OAuthClientPlainTextSecret where
+  toQueryParam = toText . unOAuthClientPlainTextSecret
+
+data OAuthClientCredentials = OAuthClientCredentials
+  { occClientId :: OAuthClientId,
+    occClientSecret :: OAuthClientPlainTextSecret
+  }
+  deriving (Eq, Show, Generic)
+  deriving (A.ToJSON, A.FromJSON, S.ToSchema) via (Schema OAuthClientCredentials)
+
+instance ToSchema OAuthClientCredentials where
+  schema =
+    object "OAuthClientCredentials" $
+      OAuthClientCredentials
+        <$> occClientId .= fieldWithDocModifier "client_id" clientIdDescription schema
+        <*> occClientSecret .= fieldWithDocModifier "client_secret" clientSecretDescription schema
+    where
+      clientIdDescription = description ?~ "The ID of the application."
+      clientSecretDescription = description ?~ "The secret of the application."
+
+data OAuthClient = OAuthClient
+  { ocId :: OAuthClientId,
+    ocName :: OAuthApplicationName,
+    ocRedirectUrl :: RedirectUrl
+  }
+  deriving (Eq, Show, Generic)
+  deriving (A.ToJSON, A.FromJSON, S.ToSchema) via (Schema OAuthClient)
+
+instance ToSchema OAuthClient where
+  schema =
+    object "OAuthClient" $
+      OAuthClient
+        <$> ocId .= field "client_id" schema
+        <*> ocName .= field "application_name" schema
+        <*> ocRedirectUrl .= field "redirect_url" schema
+
+data OAuthResponseType = OAuthResponseTypeCode
+  deriving (Eq, Show, Generic)
+  deriving (A.ToJSON, A.FromJSON, S.ToSchema) via (Schema OAuthResponseType)
+
+instance ToSchema OAuthResponseType where
+  schema :: ValueSchema NamedSwaggerDoc OAuthResponseType
+  schema =
+    enum @Text "OAuthResponseType" $
+      mconcat
+        [ element "code" OAuthResponseTypeCode
+        ]
+
+-- | The OAuth scopes that are supported by the backend.
+-- This type is a bit redundant and unfortunately has to be kept in sync
+-- with the supported scopes defined in the nginx configs.
+-- However, having this typed makes it easier to handle scopes in the backend,
+-- and e.g. provide more meaningful error messages when the scope is invalid.
```

Comment:
why isn't the scope validated and eliminated in nginz?  what can the backend not do if it doesn't know anything about scopes?  i think the answer has something to do with claim sets.  i'll keep reading, but if you have an easy explanation it might help!

Reply:
If backend doesn't know about scopes, it will just accept any text/string here. Having this type makes it much more restricted as the input it gets is parsed and if the scope is not known or formatted invalidly, it can reject the request with a proper error message. The problem with this approach, though, is that adding/changing scopes will need a new backend deployment, but this is the case anyway because the scopes are defined in the nginx.config. However, the danger is that we add a new scope but do not update this type. I left it here, because it was left over from the previous and outdated design. But I am ok with removing this type, too.